azureblob: add managed identity support

backerman opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

Adds a configuration option (`azureblob_use_msi`) to use a managed service identity when authenticating to Azure from an Azure virtual machine, which eliminates the need to manage explicit credentials. This method acquires a bearer token (valid for 24 hours) from the [Azure Instance Metadata Service](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http), and if a temporary error is returned will retry using the exponential backoff method specified in the linked documentation.

The token is not presently persisted to disk; if users are expected to invoke the `rclone` binary a large number of times in a short period that functionality would be desirable. The present implementation also does not check token validity before calling the Azure Storage blob API, so if a `rclone` invocation lasts long enough that it calls the API more than 24 hours after process start that call will fail.

#### Was the change discussed in an issue or in the forum before?

Issue #3213.

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [X] I have added tests for all changes in this PR if appropriate.
- [X] I have added documentation for the changes if appropriate.
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [X] I'm done, this Pull Request is ready for review :-)

backerman commented:
Assuming the checks don't uncover any new issues that need fixing, I'll be able to check the "ready for review" box once I've added some tests to validate that the expected errors are returned when the IMDS response is 400/500-series.

ncw commented on backend/azureblob/imds.go:
```diff
@@ -0,0 +1,141 @@
+	httpClient := fshttp.NewClient(ctx)
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		fs.Errorf(nil, "MSI is not enabled on this VM: %v", err)
```
That would probably be better returned as `errors.Wrap(err, "MSI is not enabled on this VM")` and the log isn't necessary in that case

backerman replied:
Changed these ones as suggested.

ncw commented on backend/azureblob/imds.go:
```diff
@@ -0,0 +1,141 @@
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		fs.Errorf(nil, "Couldn't read IMDS response: %v", err)
+		return result, err
```
This should be `errors.Wrap()` as well without the log.

ncw commented on backend/azureblob/imds.go:
```diff
@@ -0,0 +1,141 @@
+	// storage API call.
+	err = json.Unmarshal(b, &result)
+	if err != nil {
+		fs.Errorf(nil, "Couldn't unmarshal IMDS response: %v", err)
```
...and here

ncw reviewed (commented):
This is very nice code - thank you :-)

> Adds a configuration option (`azureblob_use_msi`) to use a managed service identity when authenticating to Azure from an Azure virtual machine, which eliminates the need to manage explicit credentials. This method acquires a bearer token (valid for 24 hours) from the [Azure Instance Metadata Service](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http), and if a temporary error is returned will retry using the exponential backoff method specified in the linked documentation.

Great

> The token is not presently persisted to disk; if users are expected to invoke the `rclone` binary a large number of times in a short period that functionality would be desirable.

Noted. If rclone users want to do this then they would be better off using the rc API.

> The present implementation also does not check token validity before calling the Azure Storage blob API, so if a `rclone` invocation lasts long enough that it calls the API more than 24 hours after process start that call will fail.

Hmm, I can see this being a potential problem. Users often kick off rclone jobs to copy TBs of data.

How difficult would it be to renew the token every few hours?

We'd then need locking around a lot more stuff though which would be rather annoying. If we want to do it then a different PR would probably be needed.

See inline for some more comments.

backerman commented on backend/azureblob/azureblob.go:
```diff
@@ -453,16 +488,63 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
-		credential, err := azblob.NewSharedKeyCredential(opt.Account, opt.Key)
+	case opt.UseMSI:
+		var token adal.Token
+		trimmedClientID := strings.TrimSpace(opt.MSIClientID)
```
Fixed.

backerman commented on backend/azureblob/imds.go:
```diff
@@ -0,0 +1,141 @@
+	// Attempt to get an MSI token; silently continue if unsuccessful.
+	// Future work: Allow user to assert that an MSI token must be
+	// available.
+	// This code has been lovingly stolen from azcopy's OAuthTokenManager.
```
I'll look around again, but I don't think so - the sample Go for getting the token does it with `net/http` as well.

backerman replied:
Looks like there is in fact `auth.NewAuthorizerFromEnvironment` which is the droid we're looking for.

backerman replied:
And that's actually only for the main Azure Go SDK; the blob storage SDK doesn't use a compatible interface.

backerman commented:
> How difficult would it be to renew the token every few hours?

Looks like it's just including a [`TokenRefresher`](https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#TokenRefresher) function where I've currently got a `nil`, so I'll do that. Shouldn't be too difficult and the pipeline package is already dealing with the concurrency issues there.

backerman commented on backend/azureblob/azureblob.go:
```diff
@@ -453,16 +488,60 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
 			return nil, errors.Wrap(err, "failed to make azure storage url from account and endpoint")
 		}
+
+		credential := azblob.NewTokenCredential(token.AccessToken, nil)
```
This is where the `TokenRefresher` goes.

backerman commented:
Okay, added token refresh support, but I'm not sure how to fix this conflict warning; there doesn't even appear to be a conflicta.

ncw commented:
Sorry for the delay responding. 

This looks ready to merge now - would you agree?

Thanks

Nick

backerman commented:
> This looks ready to merge now - would you agree?

I just rebased to master and modified the message generated when no authentication type is available (getting a bit long to list them individually). Looks good to me once the checks pass.

ncw commented:
That looks great now - thank you for the great feature :-)

I'll merge this now.

ncw commented:
Hi @backerman 

I just fixed a small oopsie in this code that the integration tests threw up cb97c2b0d39b38e1ae4f424a62503717495fa36c - that isn't a problem - these things happen!

What I was wondering about while looking at that was this in the new code you added. `opt.Account` and `opt.Endpoint` probably won't be set here will they? So maybe we don't need to make `u` at all.

https://github.com/rclone/rclone/blob/cb97c2b0d39b38e1ae4f424a62503717495fa36c/backend/azureblob/azureblob.go#L617-L620

We use it here - maybe we can pass in empty string or nil?

https://github.com/rclone/rclone/blob/cb97c2b0d39b38e1ae4f424a62503717495fa36c/backend/azureblob/azureblob.go#L645

What do you think?

backerman commented:
> Hi @backerman
> 
> I just fixed a small oopsie in this code that the integration tests threw up [cb97c2b](https://github.com/rclone/rclone/commit/cb97c2b0d39b38e1ae4f424a62503717495fa36c) - that isn't a problem - these things happen!
> 
> What I was wondering about while looking at that was this in the new code you added. `opt.Account` and `opt.Endpoint` probably won't be set here will they? So maybe we don't need to make `u` at all.

They should be set (at least `Account`) in the configuration for any authentication method; it would probably be a good idea to extract them from the switch statement, though.

ncw commented:
> > I just fixed a small oopsie in this code that the integration tests threw up [cb97c2b](https://github.com/rclone/rclone/commit/cb97c2b0d39b38e1ae4f424a62503717495fa36c) - that isn't a problem - these things happen!
> > What I was wondering about while looking at that was this in the new code you added. `opt.Account` and `opt.Endpoint` probably won't be set here will they? So maybe we don't need to make `u` at all.
> 
> They should be set (at least `Account`) in the configuration for any authentication method; it would probably be a good idea to extract them from the switch statement, though.

Ah OK!

I'm happy to leave it how it is then. Factoring it would be nice but not all the if branches need it so it probably isn't worth it.
